src/tree.rs

```diff
@@ -548,39 +627,79 @@ where
         nibble_iter: &mut NibbleIterator,
         value: Option<ValueHash>,
         tree_cache: &mut TreeCache<R>,
-    ) -> Result<PutResult<(NodeKey, Node)>> {
+        with_proof: bool,
+    ) -> Result<(PutResult<(NodeKey, Node)>, Option<SparseMerkleProof<H>>)> {
         // Find the next node to visit following the next nibble as index.
         let child_index = nibble_iter.next().expect("Ran out of nibbles");
 
         // Traverse downwards from this internal node recursively to get the `node_key` of the child
         // node at `child_index`.
-        let result = match internal_node.child(child_index) {
-            Some(child) => {
-                let child_node_key = node_key.gen_child_node_key(child.version, child_index);
-                self.insert_at(child_node_key, version, nibble_iter, value, tree_cache)?
+        let result = match internal_node.get_child_with_siblings(&node_key, child_index) {
+            (Some(child_node_key), mut siblings) => {
+                let (update_result, proof_opt) = self.insert_at(
+                    child_node_key,
+                    version,
+                    nibble_iter,
+                    value,
+                    tree_cache,
+                    with_proof,
+                )?;
+
+                let new_proof_opt = proof_opt.map(|proof| {
+                    // The move siblings function allows zero copy moves for proof
+                    let proof_leaf = proof.leaf();
+                    let mut new_siblings = proof.move_siblings();
+                    new_siblings.append(&mut siblings);
+                    SparseMerkleProof::new(proof_leaf, new_siblings)
```

Comment:
(Question) Is there a reason we create a new `SparseMerkleProof` here instead of just appending to `siblings` pf the existing proof?

Reply:
A sparse merkle proof is not (and I think shouldn't be) mutable so I decided to create a new one